apps/emqx_authn/src/emqx_authn.erl

```diff
@@ -0,0 +1,487 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2021 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_authn).
+
+-include("emqx_authn.hrl").
+
+-export([ enable/0
+        , disable/0
+        ]).
+
+-export([authenticate/1]).
+
+-export([ create_chain/1
+        , delete_chain/1
+        , lookup_chain/1
+        , list_chains/0
+        , bind/2
+        , unbind/2
+        , list_bindings/1
+        , list_bound_chains/1
+        , create_authenticator/2
+        , delete_authenticator/2
+        , update_authenticator/3
+        , lookup_authenticator/2
+        , list_authenticators/1
+        , move_authenticator_to_the_front/2
+        , move_authenticator_to_the_end/2
+        , move_authenticator_to_the_nth/3
+        ]).
+
+-export([ import_users/3
+        , add_user/3
+        , delete_user/3
+        , update_user/4
+        , lookup_user/3
+        , list_users/2
+        ]).
+
+-export([mnesia/1]).
+
+-boot_mnesia({mnesia, [boot]}).
+
+-define(CHAIN_TAB, emqx_authn_chain).
+-define(BINDING_TAB, emqx_authn_binding).
+
+%%------------------------------------------------------------------------------
+%% Mnesia bootstrap
+%%------------------------------------------------------------------------------
+
+%% @doc Create or replicate tables.
+-spec(mnesia(boot) -> ok).
+mnesia(boot) ->
```

Comment:
Soon I will merge a PR that will forbid usage of certain mnesia APIs. Details are here: https://github.com/emqx/ekka/blob/master/doc/rlog.md#enabling-rlog-in-your-application

It will be necessary to create an AuthN shard. This PR can be used as an example: https://github.com/emqx/emqx/commit/2bea95ff6e7797e428c1143f064600bdddb453e1

Reply:
I saw your changes in [emqx#5100](https://github.com/emqx/emqx/pull/5100), and I will sync them.